apps/rush-lib/src/scripts/install-run.ts

```diff
@@ -327,6 +327,10 @@ function _cleanInstallFolder(rushTempFolder: string, packageInstallFolder: strin
         'rush-recycler',
         `install-run-${Date.now().toString()}`
```

Comment:
Can we use _ensureAndJoinPath only up to 'rush-recycler', and then append the timestamped folder to this path before we run renameSync so that we don't have to create a directory and immediately remove it?

Reply:
Either is fine with me, as long as it is verified to work on Windows.  I just made the simplest fix to the existing code.

Reply:
Feel free to push commits to my branch if you want to change this, or create a new PR in place of this one.

Reply:
@mikeharder To avoid the performance cost of deleting the `node_modules` folder, the idea here is to move the folder under `common/temp/rush-recycler` which is a place that will get cleaned up at some point by the real Rush (as part of some other unrelated later operation).

The `install-run-${Date.now().toString()}` filename is trying to create a unique filename. But for some reason, on your CI system the same integer is getting generated a second time.  I wonder how that is possible.  Clock skew?

Some other ideas:

- Figure out how why this seemingly highly improbable situation is happening and fix that
- If the name already exists, roll the dice again until we get one that doesn't
- Find a different random number dispenser that is truly unique,  e.g. the last modification time of the folder + 1 is very unlikely to match an object that was already written to that folder

Alternatively, if we go with `fs.rmdirSync(rushRecyclerFolder);` we should only do this if it exists, and print a diagnostic message to the console log, because it can create other issues:

- it may be very slow
- the remove+recreate is likely to encounter race conditions on Windows, due to the way the antivirus keeps open handles to objects for a brief period after they are deleted
- `fs.rmdirSync()` will be brittle compared to the real `FileSystem.deleteFolder()` which implements elaborate retry logic

These issues are probably negligible in practice, but we should let people know. Plus we were so sure that the original thing had a negligible chance of issues. 😁

Reply:
@octogonz: I don't think there is any collision with the timestamped folder, this is just a simple bug where Windows does not allow renaming a folder to an existing folder (but Linux and Mac do allow it).  This code works on Linux/Mac but fails on Windows (Node 12):

```js
const fs = require("fs");
fs.mkdirSync("source");
fs.mkdirSync("dest");
fs.renameSync("source", "dest");
```

You should be able to repro this (and the Rush bug) on any Windows machine, it's not specific to our CI agents.

The current code in Rush first creates the timestamped folder, and then tries renaming an existing folder to the timestamped folder just created, which fails on Windows.  I see 3 solutions:

1. Delete the timestamped folder immediately after creating it.  As @halfnibble said this is both inefficient (though likely negligible to the overall operation) and confusing, but it was the smallest change to the existing code which is why I proposed it.

```ts
const rushRecyclerFolder: string = _ensureAndJoinPath(
  rushTempFolder,
  'rush-recycler',
  `install-run-${Date.now().toString()}`
);

// Windows does not allow renaming a directory to an existing directory
fs.rmdirSync(rushRecyclerFolder);

fs.renameSync(nodeModulesFolder, rushRecyclerFolder);
```

2. As @halfnibble suggested, create the path up to the `rush-recycler` folder itself, but do not create the timestamped folder.  It will be "created" automatically by the call to `renameSync`, and it should work on all platforms.  Something like this:

```ts
const rushRecyclerFolder: string = _ensureAndJoinPath(
  rushTempFolder,
  'rush-recycler'
);

const timestampedFolder = path.join(rushRecyclerFolder, `install-run-${Date.now().toString()}`);

fs.renameSync(nodeModulesFolder, timestampedFolder);
```

3. Create the timestamped folder, but then rename to a sub-folder under it:

```ts
const rushRecyclerFolder: string = _ensureAndJoinPath(
  rushTempFolder,
  'rush-recycler',
  `install-run-${Date.now().toString()}`
);

fs.renameSync(nodeModulesFolder, path.join(rushRecyclerFolder, "temp"));
```

I agree that option 2 is probably the cleanest, so I'm happy to change my PR to option 2 if you want.

Reply:
I have updated my PR to option 2.

Reply:
This looks good to me. Thank you for your input.

Reply:
😂 I am quite sure that `fs.renameSync(nodeModulesFolder, rushRecyclerFolder);` was originally intended to move `node_modules` under `rushRecyclerFolder` rather than renaming to be the directory that we just created. Given that the lab machines are all Windows boxes, I'm amazed this went unnoticed for so long.

Anyway the new code is even better because it's one level further away from `MAX_PATH`. Thanks for taking the time to fix this guys!

Reply:
@octogonz: The new code does not change the behavior in any way (other than making it work on Windows).  It is not one level further away from `MAX_PATH`.  The `node_modules` folder is still being renamed to the timestamped directory, it's just that the directory is created as part of the rename (which is guaranteed to work on all platforms), rather than being created before the rename (which fails on Windows).

Regarding why this went unnoticed for so long, it only happens when Rush needs to **re-install** itself, not the initial install.  In the real world, the most common way to hit this bug would be changing your Node version after Rush was already installed.